patch/driver-support-optoe-dynamic-write-timeout.patch

```diff
@@ -0,0 +1,129 @@
+The overwriting of the write_timeout value can be done based on platform or vendor of the transceiver.
+Following sysfs will be created with this change
+/sys/bus/i2c/devices/<bus_id>-0050/write_timeout
+
+Testing
+root@sonic:/sys/bus/i2c/devices/31-0050# ls
+dev_class  driver  eeprom  modalias  name  port_name  power  subsystem  uevent  write_max  write_timeout
+root@sonic:/sys/bus/i2c/devices/31-0050#
+
+Signed-off-by: Mihir Patel <patelmi@microsoft.com>
+---
+ ...driver-support-dynamic-write-timeout.patch | 105 ++++++++++++++++++
+ 1 file changed, 105 insertions(+)
+
+diff --git a/drivers/misc/eeprom/optoe.c b/drivers/misc/eeprom/optoe.c
+index 394c19944..a54a9ee1b 100644
+--- a/drivers/misc/eeprom/optoe.c
++++ b/drivers/misc/eeprom/optoe.c
+@@ -194,6 +194,7 @@ struct optoe_data {
+ 
+ 	u8 *writebuf;
+ 	unsigned int write_max;
++	unsigned int write_timeout;
+ 
+ 	unsigned int num_addresses;
+ 
+@@ -223,7 +224,8 @@ static unsigned int io_limit = OPTOE_PAGE_SIZE;
+  * specs often allow 5 msec for a page write, sometimes 20 msec;
+  * it's important to recover from write timeouts.
+  */
+-static unsigned int write_timeout = 25;
++#define OPTOE_DEFAULT_WRITE_TIMEOUT 25
++#define OPTOE_MAX_SUPPORTED_WRITE_TIMEOUT 500
+ 
+ /*
+  * flags to distinguish one-address (QSFP family) from two-address (SFP family)
+@@ -352,7 +354,7 @@ static ssize_t optoe_eeprom_read(struct optoe_data *optoe,
+ 	 * loop a few times until this one succeeds, waiting at least
+ 	 * long enough for one entire page write to work.
+ 	 */
+-	timeout = jiffies + msecs_to_jiffies(write_timeout);
++	timeout = jiffies + msecs_to_jiffies(optoe->write_timeout);
+ 	do {
+ 		read_time = jiffies;
+ 
+@@ -453,7 +455,7 @@ static ssize_t optoe_eeprom_write(struct optoe_data *optoe,
+ 	 * loop a few times until this one succeeds, waiting at least
+ 	 * long enough for one entire page write to work.
+ 	 */
+-	timeout = jiffies + msecs_to_jiffies(write_timeout);
++	timeout = jiffies + msecs_to_jiffies(optoe->write_timeout);
+ 	do {
+ 		write_time = jiffies;
+ 
+@@ -855,6 +857,39 @@ static ssize_t set_dev_write_max_size(struct device *dev,
+ 	return count;
+ }
+ 
++static ssize_t show_dev_write_timeout_size(struct device *dev,
++			struct device_attribute *dattr, char *buf)
++{
++	struct i2c_client *client = to_i2c_client(dev);
++	struct optoe_data *optoe = i2c_get_clientdata(client);
++	ssize_t count;
++
++	mutex_lock(&optoe->lock);
++	count = sprintf(buf, "%u\n", optoe->write_timeout);
++	mutex_unlock(&optoe->lock);
++
++	return count;
++}
++
++static ssize_t set_dev_write_timeout_size(struct device *dev,
++			struct device_attribute *attr,
++			const char *buf, size_t count)
++{
++	struct i2c_client *client = to_i2c_client(dev);
++	struct optoe_data *optoe = i2c_get_clientdata(client);
++	unsigned int write_timeout_size;
++
++	if (kstrtouint(buf, 0, &write_timeout_size) != 0 ||
++		write_timeout_size < 0 || write_timeout_size > OPTOE_MAX_SUPPORTED_WRITE_TIMEOUT)
```

Comment:
Considering `write_timeout_size` is an unsigned int, `write_timeout_size < 0` will (likely) never be true.

As an alternative, is it better to have a required minimum of 5ms or 20ms?

Reply:
Agreed that `write_timeout_size < 0` will likely not be true.
I have currently kept 0ms as minimum timeout value since didn't want to have a pre-defined minimum value which can potentially prevent future wait_timeout optimization. This would be the case if a transceiver in future supports wait_timeout less than pre-defined minimum timeout (although less likely)